runtime/secrets/secrets.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2025 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secrets
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// TLSCertKey is the standard key for TLS certificate data in secrets.
+	TLSCertKey = corev1.TLSCertKey
+	// TLSKeyKey is the standard key for TLS private key data in secrets.
+	TLSKeyKey = corev1.TLSPrivateKeyKey
+	// CACertKey is the standard key for CA certificate data in secrets.
+	CACertKey = "ca.crt"
+
+	// TLSCertFileKey is the deprecated key for TLS certificate data in secrets.
+	TLSCertFileKey = "certFile"
+	// TLSKeyFileKey is the deprecated key for TLS private key data in secrets.
+	TLSKeyFileKey = "keyFile"
+	// CACertFileKey is the deprecated key for CA certificate data in secrets.
+	CACertFileKey = "caFile"
+
+	// UsernameKey is the key for username data in basic auth secrets.
+	UsernameKey = "username"
+	// PasswordKey is the key for password data in basic auth secrets.
+	PasswordKey = "password"
+
+	// ProxyAddressKey is the key for proxy address data in proxy secrets.
+	ProxyAddressKey = "address"
+
+	// BearerTokenKey is the key for bearer token data in secrets.
+	BearerTokenKey = "bearerToken"
+	// TokenKey is the key for generic API token data in secrets.
+	TokenKey = "token"
+)
+
+// tlsCertificateData holds TLS certificate, key, and optional CA data
+type tlsCertificateData struct {
+	cert   []byte
+	key    []byte
+	caCert []byte
+}
+
+// newTLSCertificateData creates tlsCertificateData from a Kubernetes secret.
+func newTLSCertificateData(secret *corev1.Secret, supportDeprecated bool) (*tlsCertificateData, error) {
+	data := &tlsCertificateData{
+		cert:   getSecretData(secret, TLSCertKey, TLSCertFileKey, supportDeprecated),
+		key:    getSecretData(secret, TLSKeyKey, TLSKeyFileKey, supportDeprecated),
+		caCert: getSecretData(secret, CACertKey, CACertFileKey, supportDeprecated),
+	}
+
+	if err := data.validate(); err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (t *tlsCertificateData) validate() error {
+	hasCert := len(t.cert) > 0
+	hasKey := len(t.key) > 0
+	hasCA := len(t.caCert) > 0
+
+	if hasCert != hasKey {
+		if hasCert {
+			return fmt.Errorf("found certificate but missing private key")
+		}
+		return fmt.Errorf("found private key but missing certificate")
+	}
+
+	if !hasCert && !hasCA {
+		return fmt.Errorf("no CA certificate or client certificate pair found")
+	}
```

Comment:
These error messages are fine for `MakeTLSSecret`, but for `TLSConfigFromSecret` we need to be specific about which keys are missing inside the secret, and the keys must be enclosed by single-quotes in the error message. Also, all the error messages for the public functions of `reader.go` must also mention the secret with the format `'<namespace>/<name>'`. This way we don't need to do this in the various places where we will call this library.

Reply:
Thanks! Please add the mention to the secret ref in the error messages of the other public functions in `reader.go` as well.